bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -585,125 +591,143 @@ private void cancelPollingJob() {
+            zoneData.setVolume(newZoneData.getVolume());
+            zoneData.setSource(newZoneData.getSource());
+            updateChannelState(zoneData.getZone(), CHANNEL_TYPE_POWER);
+            updateChannelState(zoneData.getZone(), CHANNEL_TYPE_MUTE);
+            updateChannelState(zoneData.getZone(), CHANNEL_TYPE_VOLUME);
+            updateChannelState(zoneData.getZone(), CHANNEL_TYPE_SOURCE);
 
-                if (!matcher.group(11).equals(zoneData.getKeypad())) {
-                    zoneData.setKeypad(matcher.group(11));
-                    updateChannelState(zone, CHANNEL_TYPE_KEYPAD, zoneData);
-                }
-            } else {
-                logger.debug("Invalid zone update message: {}", newZoneData);
+        }
+        lastPollingUpdate = System.currentTimeMillis();
+    }
+
+    /**
+     * Update the state of a channel
+     *
+     * @param zoneId the zone id used to lookup the channel to be updated
+     * @param channelType the channel type to be updated
+     */
+    private void updateChannelState(String zoneId, String channelType) {
+        MonopriceAudioZoneDTO zoneData = zoneDataMap.get(zoneId);
+
+        if (zoneData != null) {
+            String channel = amp.getZoneName(zoneId) + CHANNEL_DELIMIT + channelType;
+
+            if (!isLinked(channel)) {
+                return;
             }
 
+            logger.debug("updating channel state for zone: {}, channel type: {}", zoneId, channelType);
+
+            State state = UnDefType.UNDEF;
+            switch (channelType) {
+                case CHANNEL_TYPE_POWER:
+                    state = zoneData.isPowerOn() ? OnOffType.ON : OnOffType.OFF;
+                    break;
+                case CHANNEL_TYPE_SOURCE:
+                    state = new DecimalType(zoneData.getSource());
+                    break;
+                case CHANNEL_TYPE_VOLUME:
+                    long volumePct = Math.round(
+                            (zoneData.getVolume() - MIN_VOLUME) / (double) (amp.getMaxVol() - MIN_VOLUME) * 100.0);
+                    state = new PercentType(BigDecimal.valueOf(volumePct));
+                    break;
+                case CHANNEL_TYPE_MUTE:
+                    state = zoneData.isMuted() ? OnOffType.ON : OnOffType.OFF;
+                    break;
+                case CHANNEL_TYPE_TREBLE:
+                    state = new DecimalType(BigDecimal.valueOf(zoneData.getTreble() - amp.getToneOffset()));
+                    break;
+                case CHANNEL_TYPE_BASS:
+                    state = new DecimalType(BigDecimal.valueOf(zoneData.getBass() - amp.getToneOffset()));
+                    break;
+                case CHANNEL_TYPE_BALANCE:
+                    state = new DecimalType(BigDecimal.valueOf(zoneData.getBalance() - amp.getBalOffset()));
+                    break;
+                case CHANNEL_TYPE_DND:
+                    state = zoneData.isDndOn() ? OnOffType.ON : OnOffType.OFF;
```

Comment:
Could also be written as:
```suggestion
                    state = OnOffType.from(zoneData.isDndOn());
```

Reply:
That is cleaner!
All comments should now be addressed.